### ISSUE TITLE
Does not build for node 7.10.1 WeakCallbackData does not exist

### ISSUE THREAD

**Issue by rwlaschin** [51Degrees/Device-Detection#16]
When attempting to install with Node 7.10.1  the compile step fails due to missing template WeakCallbackData (see below).  Node 7 > has changed the api and deprecated some of the modules and classes.

It would be great to get an update so this will work.

Reproduction
- Install node 7.10.1
- npm install fiftyonedegreescore 

Here are some of the errors : 
../src/pattern/51Degrees_node.cpp:966:27: error: no template named 'WeakCallbackData' in namespace 'v8'; did you mean 'WeakCallbackInfo'?
  void (*dtor) (const v8::WeakCallbackData<v8::Object, SWIGV8_Proxy> &data);
                      ~~~~^~~~~~~~~~~~~~~~
                          WeakCallbackInfo
/Users/robert/.node-gyp/7.10.1/include/node/v8.h:7475:16: note: 'WeakCallbackInfo' declared here
  friend class WeakCallbackInfo;
               ^
../src/pattern/51Degrees_node.cpp:966:27: error: too many template arguments for class template 'WeakCallbackInfo'
  void (*dtor) (const v8::WeakCallbackData<v8::Object, SWIGV8_Proxy> &data);
                          ^                            ~~~~~~~~~~~~~
/Users/robert/.node-gyp/7.10.1/include/node/v8.h:7475:16: note: template is declared here
  friend class WeakCallbackInfo;
               ^
../src/pattern/51Degrees_node.cpp:1014:53: error: no template named 'WeakCallbackData' in namespace 'v8'; did you mean 'WeakCallbackInfo'?
SWIGRUNTIME void SWIGV8_Proxy_DefaultDtor(const v8::WeakCallbackData<v8::Object, SWIGV8_Proxy> &data) {
                                                ~~~~^~~~~~~~~~~~~~~~
                                                    WeakCallbackInfo

**Comment by Joseph51D** [comment]
Hi @rwlaschin 

Thank you for the information. Currently our node developer is away on leave but I'll make sure to get this to him on his return next week.

Regards,
Joe

**Comment by ben51degrees** [comment]
@rwlaschin The 51Degrees_node.cpp file is generated by SWIG, which currently does not support node 7 as stable. There is however work going on with this in [this pull request](https://github.com/swig/swig/pull/968) which has made a lot of progress and currently passes all the SWIG CI tests. We cannot introduce this to our source as stable until the SWIG feature is considered stable first, although all the tests I have run on our API with node 7 have not brought up any problems which are immediately apparent.

In the absence of a stable release, I have attached the latest NPM package with the necessary change. The  src/pattern/51Degrees_node.cpp file has been replaced with one generated using SWIG built from the above pull request. This works fine for me, although if there are any issues you discover, do let me know and we can look into it.

After some more thorough testing, the aim will be to release this as a beta to make installation easier in the mean time. I will keep you updated on when we intend to release this, and subsequently a stable package which supports node 7.

[fiftyonedegreescore-3.2.16.zip](https://github.com/51Degrees/Device-Detection/files/1231751/fiftyonedegreescore-3.2.16.zip) (unzip first)

Regards,
Ben

**Comment by Osterjour** [comment]
Hey Ben,
just to make sure you're aware of the fact that we're waiting for Node.js 8 support too.
To be honest I don't get why this has to come up in August 2017 as Node v7 is available since October 2016 and Node.js 8 is the stable LTS version since May 2017.
Unfortunately you don't mention the fact that you're not supporting the current stable release of Node.js anywhere. So during our upgrade research we assumed the package will just work :(
I understand that you rely on dependencies here, but you should at least push those dependencies as soon as possible (which is October 2016) and additionally inform your customers about the missing support for all versions >6.

Regards,
Markus

**Comment by ben51degrees** [comment]
@Osterjour 
I can confirm the patch I supplied above supports Node v8 too. The only issue was the deprecation of WeakCallbackData, which means version >v7 are supported by the patch.

Thanks for raising the issue that the supported version is not stated. We will be releasing the patched version to beta in the next few weeks, to both GitHub and NPM. So in the same release we will add a statement about versions, and version lock the NPM packages.

Regards,
Ben.

**Comment by ashishnetworks** [comment]
can somebody please confirm when will the support for Node version 8.0.0 start? OR at least tell me way/patch to make current version work with Node 8.0

**Comment by ben51degrees** [comment]
@ashishnetworks 
Your timing is very fortuitous. I've just this minute released the changes. The dev/node-7-up branch supports versions 7 and above as of commit 6625e77.

It has also been released to NPM as version 3.2.18-dev under the tag "dev", so the quickest way to install this is with `npm install fiftyonedegreescore@dev`.

For now this will remain a dev version until the SWIG version used is considered stable.

Regards,
Ben.

**Comment by ben51degrees** [comment]
Closed in version 3.2.12.12 by commit 6625e77.

**Comment by ashishnetworks** [comment]
Thanks. Above branch worked for me. I'm running on Node 8.0.0
